server/src/errors.rs

```diff
@@ -0,0 +1,41 @@
+/* Copyright 2022 The MiniModelarDB Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//! The error types used throughout the system. Their design is based on [Rust
+//! by Example], [Apache Arrow],  and [Apache Arrow DataFusion].
+//!
+//! [Rust by Example] https://doc.rust-lang.org/rust-by-example/error/multiple_error_types/define_error_type.html
+//! [Apache Arrow] https://github.com/apache/arrow-rs/blob/master/arrow/src/error.rs
+//! [Apache Arrow DataFusion] https://github.com/apache/arrow-datafusion/blob/master/datafusion/common/src/error.rs
+
+use std::error::Error;
+use std::fmt::{Display, Formatter};
+
+/// Error type used throughout the system.
+#[derive(Debug)]
+pub enum MiniModelarDBError {
+    /// Error returned by the model types.
+    CompressionError(String),
```

Comment:
Generally I think the different kinds of errors that can occur should be represented instead of where they occur, e.g., using an `IoError` instead of a `StorageEngine` error. In terms of granularity I suggest we use the error kinds as categories of errors, e.g., `IoError` and than add a structured member that indicates exactly what happened if we need additional detail. For example see [`std::io:error`](https://doc.rust-lang.org/std/io/struct.Error.html). Of course we can also add  error kinds with different granularity to the enum, but then it is not clear to me how we can model the relationship between low level and high level error kinds.

Reply:
I agree with your suggestion for granularity.